cns/kubecontroller/pod/reconciler.go

```diff
@@ -2,79 +2,99 @@ package pod
 
 import (
 	"context"
+	"strconv"
 
 	"github.com/pkg/errors"
+	"go.uber.org/zap"
 	v1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/fields"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/event"
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-type podcli interface {
+type cli interface {
 	List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error
 }
 
-type podListener interface {
-	Update([]v1.Pod)
+// watcher watches Pods on the current Node and notifies listeners of changes.
+type watcher struct {
+	z              *zap.Logger
+	cli            cli
+	reconcileFuncs []reconcile.Func
 }
 
-type PodWatcher struct {
-	cli            podcli
-	listOpt        client.ListOption
-	ReconcileFuncs []reconcile.Func
+func New(z *zap.Logger) *watcher { //nolint:revive // force usage of new by keeping the struct private
```

Comment:
I think it's fine to leave the struct exported so long as there is a good comment indicating that it should be constructed with `New`. However the flip side is that I think we should only export things when necessary, so if there is no reason to export, then it's fine like this

Reply:
Poka-yoke. If I let you construct the struct directly and you do not know that you must pass a logger reference, everything will NPE.